---
title: "mRNA differential expression in PTOA"
format: html
editor: visual
---

Load the libraries needed

```{r include=FALSE}
library(DESeq2)
library(BiocParallel)
library(ggplot2)
library(tidyverse)
library(ggpubr)
library(cowplot)
library(EnsDb.Hsapiens.v103)
library(EnsDb.Mmusculus.v103)
library(biomaRt)
library(readxl)
library(feather)
library(writexl)
library(ggsci)
library(ggsci)
library(RColorBrewer)
library(limma)
library(edgeR)
library(rcartocolor)

```

Perform differential expression analysis of the mRNA data

```{r include=FALSE}
source("src/utilityFunctions.R")

#get the OATargets data to annotate genes previously experimentally perturbed in PTOA
annotationData <- read_excel("data/AnnotationData-2022-10-27.xlsx",sheet = 2)

#get the consensus effect - change the column to consider no effect experiments to be superseded by experiments where they show an effect
annotationData <- annotationData %>% dplyr::filter(`Inferred gene effect`!= "No effect") %>% dplyr::group_by(Gene) %>%
  dplyr::mutate(effectConsensus=case_when(dplyr::n_distinct(`Inferred gene effect`)>1 ~ "Ambiguous",TRUE ~ `Inferred gene effect`)) %>%
  dplyr::filter(row_number(Gene) == 1) %>% as.data.frame()

#get a table of the human and gene symbols
human2geneTable <- convertHumanGeneList(annotationData$Gene)

#merge the annotation table to this table as OATargets is human gene name based
annotationData <- merge(annotationData,human2geneTable,by.x="Gene",by.y="HGNC.symbol")


#get the sample table, the comparisons to do and the gene x count matrix from the SkeletalVis pipeline
sampleTable <- "data/ACLModel_sampleTable.txt"
comparisonsTable <- "data/ACLModel_comparisons.txt"

#note - using length scaled count data with limma-voom as recommended in the tximport vignette
txiData  <- "data/ACLModel_txiScaledTPM.RDS"

#run the differential expression analysis pipeline using limma voom as there are pairs of samples with the same condition (naive) that cannot be analysed with DESeq2
res <- runDiffExp(txiData,sampleTable,comparisonsTable,annotationData,block="Mouse",variables = "Condition",method="limma",PCAColumns = c("Day","Treatment"),"ACLModel_mRNADiff.txt",species = "Mouse",fcThreshold = 1.5)

#number of differentially expressed genes
d7aclvsnaive <- res$cutTables[[2]]
numUp <- nrow(na.omit(d7aclvsnaive[d7aclvsnaive$logFC >= log2(1.5) & d7aclvsnaive$adj.P.Val <=0.05,]))
sprintf("number upregulated genes is %d",numUp)

numDown <- nrow(na.omit(d7aclvsnaive[d7aclvsnaive$logFC <= log2(1/1.5) & d7aclvsnaive$adj.P.Val <=0.05,]))
sprintf("number downregulated genes is %d",numDown)

#save as an excel file
cutTables <- res$cutTables
names(cutTables) <- c("D7ACLvsConlateral","D7ACLvsNaive")
write_xlsx(cutTables,"results/SupplementryTable1.xlsx")

#save the differential expression results for later use
saveRDS(res$cutTables[[2]],file="data/D7ACLvsNaive_diffExp.RDS")


```

GO analysis of the differentially expressed genes

```{r include=FALSE}
source("src/GOEnrichment.R")

#read in the gene lengths for GO enrichment
geneLengths <- read.delim("data/mouseGeneLengths.txt")

#order the results in the correct format
d7aclvsnaive <- d7aclvsnaive[,c("gene_name","logFC","adj.P.Val","ID")]

write.table(d7aclvsnaive, file = "results/d7aclvsnaive_mRNA.txt", col.names = T,row.names = F, sep = "\t", quote = F)

#perform GO enrichment
#just up genes
goresults.up <- GOEnrichment(differentialExpression = d7aclvsnaive,geneLengths = geneLengths,species="Mouse",foldChangeOnly = FALSE,foldchange = 1.5 ,padj = 0.05, direction="up")

#just down genes

goresults.down <- GOEnrichment(differentialExpression = d7aclvsnaive,geneLengths = geneLengths,species="Mouse",foldChangeOnly = FALSE,foldchange = 1.5 ,padj = 0.05, direction="down")

#save the result
write_xlsx(list(GO.Up=goresults.up[[1]],GO.Down=goresults.down[[1]]),"results/SupplementryTable2.xlsx")

#plot significant goterms of interest
IDs <- c("GO:0030198","GO:0001503","GO:0009611","GO:0007266","GO:0009612","GO:0045597","GO:0016055")
goPlots.up <- plotEnrichment(goresults.up[[1]],IDs = IDs)

IDs <- c("GO:0034976","GO:0044257","GO:0051056","GO:0070085","GO:1901565","GO:0016192")
goPlots.down <- plotEnrichment(goresults.down[[1]],IDs=IDs) + scale_x_continuous(breaks = scales::pretty_breaks(n = 4),expand = expansion(mult = c(0.15, 0.15)))

```

Compare the observed log2 fold changes against transcriptomic experiments from skeletalvis

```{r include=FALSE}
#set up the skeletalvis log2 foldchanges and metadata
accessions <- read.delim("data/accessions.txt")
accessions$combined <- paste(accessions$accession,accessions$comparison,sep="_")

foldchangeTable <- as.data.frame(read_feather("data/foldChangeTable.feather"))
rownames(foldchangeTable)<- foldchangeTable$ID

#get a table of the human and gene symbols
human2geneTable <- convertHumanGeneList(foldchangeTable$ID)

d7aclvsnaive <- merge(d7aclvsnaive,human2geneTable,by.x="gene_name",by.y="MGI.symbol")
d7aclvsnaive <- d7aclvsnaive[,c(5,2:4)]
colnames(d7aclvsnaive)[1] <- "GeneSymbol"

#get the similarity of the acl model log2 foldchanges to each dataset in skeletalvis with cosine similarity
sims <-  getSim(d7aclvsnaive,foldchangeTable,accessions,"ACL")
sims <- sims[ order(sims$zscore,decreasing =TRUE),c("accession","comparisonsText","zscore")]


write_xlsx(sims,"results/SupplementryTable3.xlsx")

simPlot <- ggplot(sims,aes(x = zscore,fill=abs(zscore)>=2)) + 
  geom_histogram(bins = 100)  +  scale_fill_carto_d( palette = "Safe") +  cowplot::theme_cowplot(font_size = 28) + xlab("cosine similarity zscore") + theme(legend.position = "none") + theme(axis.line = element_line(size = 1))

#plot the selected similar datasets that are PTOA transcriptomic experiments
skeletalVisPlot <- plotSkeletalVis(sims)


```

Save all the plots

```{r}
save_plot("figures/Fig1A_PCA.png",res$pca,base_height = 8,base_width = 8,bg="white")
save_plot("figures/Fig1B_Volcano.png",res$volcano[[2]],base_height = 6,base_width = 6,bg="white")
save_plot("figures/Fig1C_OATargets.png",res$oatargets[[2]][[1]],base_height = 6,base_width = 6,bg="white")
save_plot("figures/Fig1D_SkeletalVisSimilarity.png",skeletalVisPlot,base_height = 6,base_width = 7,bg="white")
save_plot("figures/Fig1E_GOMDS_up.png",goPlots.up,base_height = 11,base_width = 12,bg="white")
save_plot("figures/Fig1F_GOMDS_down.png",goPlots.down,base_height = 11,base_width = 13,bg="white")

```

Stats for OATarget proportion of detrimental vs protective genes

```{r}

diffTargets <- res$oatargets[[2]][[2]]
totals <- table(diffTargets$effectConsensus)[c("Protective","Detrimental")]
totals

diffTargets.degs <- diffTargets[ diffTargets$logFC >= log2(1.5) & diffTargets$adj.P.Val <= 0.05,]

totals.degs <- table(diffTargets.degs$effectConsensus)[c("Protective","Detrimental")]
totals.degs
prop.test(totals.degs, totals)
          
```
